examples/scene/volume_2dtf.py

```diff
@@ -0,0 +1,217 @@
+# vispy: gallery 2
+
+"""
+Volume Rendering with 2D Transfer Function
+================
+
+Example volume rendering with an editable 2D transfer function.
+
+The bottom view shows the 2D histogram of the data (value, gradent magnitude)
+with the transfer function overlaid.
+
+Controls:
+
+* 1  - Reset to default 1D transfer function
+* 2  - Reset to default 2D transfer function
+* 3  - Show attenuated MIP
+* 4  - Show attenuated MIP colored by depth of max value
+
+# Left click and drag to paint the transfer function
+# Alt + Left click and drag to erase on the transfer function
+
+"""
+import numpy as np
+from vispy import app, scene, io
+from vispy.color import BaseTransferFunction
+
+# Read and normalize volume data
+mri_data = np.load(io.load_data_file('brain/mri.npz'))['data']
+mri_data = np.flipud(np.rollaxis(mri_data, 1)).astype(np.float32)
+mri_data = (mri_data - mri_data.min()) / mri_data.ptp()
+
+# Prepare canvas
+canvas = scene.SceneCanvas(keys='interactive', size=(768, 1024), show=True)
+canvas.measure_fps()
+grid = canvas.central_widget.add_grid()
+
+# Set up a viewbox to display the image
+data_view = grid.add_view(row=0, col=0, row_span=2)
+# The other view will show the 2D histogram of the data with the TF over it
+cmap_view = grid.add_view(row=2, col=0, camera='panzoom')
+
+
+class GradientMagnitudeTF(BaseTransferFunction):
+    """Transfer function that uses the data value and gradient magnitude to
+    sample a 2D transfer function.
+    """
+
+    glsl_tf = """\
+    vec4 applyTransferFunction(vec4 color, vec3 loc, vec3 step) {
+        step = step / u_relative_step_size;
+        vec4 prev;
+        vec4 next;
+        // calculate normal vector from gradient
+        vec3 N; // normal
+        prev = $get_data(loc + vec3(-step.x, 0.0, 0.0));
+        next = $get_data(loc + vec3(step.x, 0.0, 0.0));
+        N[0] = colorToVal(prev) - colorToVal(next);
+        prev = $get_data(loc + vec3(0.0, -step.y, 0.0));
+        next = $get_data(loc + vec3(0.0, step.y, 0.0));
+        N[1] = colorToVal(prev) - colorToVal(next);
+        prev = $get_data(loc + vec3(0.0, 0.0, -step.z));
+        next = $get_data(loc + vec3(0.0, 0.0, step.z));
+        N[2] = colorToVal(prev) - colorToVal(next);
+        float gm = length(N); // gradient magnitude
+
+        gm = clamp(gm, min(clim.x, clim.y), max(clim.x, clim.y));
+
+        float data = colorToVal(color);
+        data = clamp(data, min(clim.x, clim.y), max(clim.x, clim.y));
+
+        vec4 final_color = sampleTFLUT(pow(data, gamma), gm);
+        return final_color;
+    }
+    """
+
+
+class DepthColorTF(BaseTransferFunction):
+    """Transfer function that uses the data value and gradient magnitude to
+    sample a 2D transfer function.
+    """
+
+    glsl_tf = """\
+    vec4 applyTransferFunction(vec4 color, vec3 loc, vec3 step) {
+        vec4 hue = applyColormap(length(loc));
+        vec4 final_color = vec4(color.r * hue.rgb, 1.0);
+        return final_color;
+    }
+    """
```

Comment:
@royerloic will love this, we couldn't figure it out last time but we just needed to use `loc` :P

Reply:
Initially I was just working on the 2D LUT stuff, but was inspired by a conversation with him to add this feature. It was very simple on this branch, which felt like a good sign! Note this is what I called out in the PR description though - in the MIP shader it's actually passing in `max_loc_tex - start_loc` for this parameter. I'm not sure if that's the best though. An alternative would be to pass in `start_loc` separately, but I also don't want to pass in a bunch of generally unused parameters.

Reply:
I see... well, I think it's fair to pass as many parameters as we think might be useful :shrug: It's easy to not use them.

Alternatively, glsl simply accessing stuff from the outer scope generally works (though it's uglier).